File: 11.camera_display/python/camera_display/camera_display_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Camera Display Node (Python version)

Simultaneously displays RGB and depth images from both head and waist cameras.
Supports depth colormap, histogram, and statistics overlay.
"""

import threading
from dataclasses import dataclass, field
from typing import Optional

import cv2
import numpy as np

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy
from sensor_msgs.msg import Image


@dataclass
class CameraData:
    """Holds image data and statistics for one camera."""
    label: str
    depth_image: Optional[np.ndarray] = None
    colored_depth: Optional[np.ndarray] = None
    color_image: Optional[np.ndarray] = None
    min_depth_value: int = 0
    max_depth_value: int = 0
    mean_depth_value: float = 0.0
    valid_pixel_count: int = 0
    lock: threading.Lock = field(default_factory=threading.Lock)


class CameraDisplayNode(Node):
    """ROS2 Node for dual-camera (head + waist) image display"""

    COLORMAP_CODES = {
        0: None,  # GRAY
        1: cv2.COLORMAP_JET,
        2: cv2.COLORMAP_RAINBOW,
        3: cv2.COLORMAP_TURBO,
    }

    HEAD_NS = 'ob_camera_head'
    WAIST_NS = 'ob_camera_waist'

    def __init__(self):
        super().__init__('camera_display_node')

        # Declare parameters
        self.declare_parameter('colormap', 2)
        self.declare_parameter('max_depth', 5000.0)
        self.declare_parameter('min_depth', 0.0)
        self.declare_parameter('display_scale', 0.5)
        self.declare_parameter('show_histogram', True)
        self.declare_parameter('show_statistics', True)
        self.declare_parameter('enable_head', True)
        self.declare_parameter('enable_waist', True)

        self.colormap = self.get_parameter('colormap').get_parameter_value().integer_value
        self.max_depth = self.get_parameter('max_depth').get_parameter_value().double_value
        self.min_depth = self.get_parameter('min_depth').get_parameter_value().double_value
        self.display_scale = self.get_parameter('display_scale').get_parameter_value().double_value
        self.show_histogram = self.get_parameter('show_histogram').get_parameter_value().bool_value
        self.show_statistics = self.get_parameter('show_statistics').get_parameter_value().bool_value
        self.enable_head = self.get_parameter('enable_head').get_parameter_value().bool_value
        self.enable_waist = self.get_parameter('enable_waist').get_parameter_value().bool_value

        qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )

        self.cameras: dict[str, CameraData] = {}

        if self.enable_head:
            self._setup_camera(self.HEAD_NS, 'Head', qos)
        if self.enable_waist:
            self._setup_camera(self.WAIST_NS, 'Waist', qos)

        if not self.cameras:
            self.get_logger().error('No camera enabled! Set enable_head or enable_waist to true.')
            return

        self.display_running = True
        self.display_thread = threading.Thread(target=self.display_loop, daemon=True)
        self.display_thread.start()

        self.get_logger().info('Camera Display Node initialized')
        for ns, cam in self.cameras.items():
            self.get_logger().info(f'  {cam.label} camera:')
            self.get_logger().info(f'    Depth: {ns}/depth/image_raw')
            self.get_logger().info(f'    Color: {ns}/color/image_raw')
        self.get_logger().info(f'  Colormap: {self.colormap}')
        self.get_logger().info(f'  Depth range: {self.min_depth:.0f} - {self.max_depth:.0f} mm')

    def _setup_camera(self, namespace, label, qos):
        cam = CameraData(label=label)
        self.cameras[namespace] = cam

        self.create_subscription(
            Image, f'{namespace}/depth/image_raw',
            lambda msg, c=cam: self._depth_callback(msg, c), qos
        )
        self.create_subscription(
            Image, f'{namespace}/color/image_raw',
            lambda msg, c=cam: self._color_callback(msg, c), qos
        )

    def _depth_callback(self, msg, cam: CameraData):
        try:
            if msg.encoding in ('16UC1', 'mono16'):
                depth = np.frombuffer(msg.data, dtype=np.uint16).reshape(msg.height, msg.width)
            elif msg.encoding == '32FC1':
                depth_float = np.frombuffer(msg.data, dtype=np.float32).reshape(msg.height, msg.width)
                depth = (depth_float * 1000).astype(np.uint16)
            else:
                self.get_logger().warn(f'Unsupported depth encoding: {msg.encoding}')
                return

            with cam.lock:
                cam.depth_image = depth.copy()
                self._update_depth_statistics(depth, cam)
                cam.colored_depth = self._apply_colormap(depth)

        except Exception as e:
            self.get_logger().error(f'Error processing depth image ({cam.label}): {e}')

    def _color_callback(self, msg, cam: CameraData):
        try:
            if msg.encoding in ('rgb8', 'RGB8'):
                color = np.frombuffer(msg.data, dtype=np.uint8).reshape(msg.height, msg.width, 3)
                color = cv2.cvtColor(color, cv2.COLOR_RGB2BGR)
            elif msg.encoding in ('bgr8', 'BGR8'):
                color = np.frombuffer(msg.data, dtype=np.uint8).reshape(msg.height, msg.width, 3)
            elif msg.encoding in ('rgba8', 'RGBA8'):
                color = np.frombuffer(msg.data, dtype=np.uint8).reshape(msg.height, msg.width, 4)
                color = cv2.cvtColor(color, cv2.COLOR_RGBA2BGR)
            elif msg.encoding in ('bgra8', 'BGRA8'):
                color = np.frombuffer(msg.data, dtype=np.uint8).reshape(msg.height, msg.width, 4)
                color = cv2.cvtColor(color, cv2.COLOR_BGRA2BGR)
            else:
                self.get_logger().warn(f'Unsupported color encoding: {msg.encoding}')
                return

            with cam.lock:
                cam.color_image = color.copy()

        except Exception as e:
            self.get_logger().error(f'Error processing color image ({cam.label}): {e}')

    @staticmethod
    def _update_depth_statistics(depth, cam: CameraData):
        mask = depth > 0
        if not np.any(mask):
            cam.min_depth_value = 0
            cam.max_depth_value = 0
            cam.mean_depth_value = 0.0
            cam.valid_pixel_count = 0
            return
        cam.min_depth_value = int(np.min(depth[mask]))
        cam.max_depth_value = int(np.max(depth[mask]))
        cam.mean_depth_value = float(np.mean(depth[mask]))
        cam.valid_pixel_count = int(np.sum(mask))

    def _apply_colormap(self, depth):
        normalized = np.clip(
            (depth - self.min_depth) * 255 / (self.max_depth - self.min_depth),
            0, 255
        ).astype(np.uint8)

        colormap_code = self.COLORMAP_CODES.get(self.colormap, cv2.COLORMAP_JET)
        if colormap_code is None:
            colored = cv2.cvtColor(normalized, cv2.COLOR_GRAY2BGR)
        else:
            colored = cv2.applyColorMap(normalized, colormap_code)

        colored[depth == 0] = [0, 0, 0]
        return colored

    def _create_histogram(self, depth):
        mask = depth > 0
        if not np.any(mask):
            return None

        hist = cv2.calcHist([depth], [0], mask.astype(np.uint8), [256],
                            [self.min_depth, self.max_depth])
        cv2.normalize(hist, hist, 0, 1, cv2.NORM_MINMAX)

        hist_h, hist_w = 200, 512
        hist_image = np.zeros((hist_h, hist_w, 3), dtype=np.uint8)
        bin_w = hist_w // 256

        for i in range(1, 256):
            cv2.line(
                hist_image,
                (bin_w * (i - 1), hist_h - int(hist[i - 1] * hist_h)),
                (bin_w * i, hist_h - int(hist[i] * hist_h)),
                (255, 255, 255), 2
            )

        label = f'Range: {int(self.min_depth)}-{int(self.max_depth)} mm'
        cv2.putText(hist_image, label, (10, 20),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
        return hist_image

    def display_loop(self):
        # Create windows for enabled cameras
        for cam in self.cameras.values():
            cv2.namedWindow(f'{cam.label} - Depth', cv2.WINDOW_NORMAL)
            cv2.namedWindow(f'{cam.label} - Color', cv2.WINDOW_NORMAL)

        while self.display_running and rclpy.ok():
            for cam in self.cameras.values():
                with cam.lock:
                    if cam.colored_depth is not None:
                        depth_display = cam.colored_depth.copy()

                        if self.show_statistics:
                            cv2.putText(depth_display, f'Min: {cam.min_depth_value} mm',
                                        (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (255, 255, 255), 2)
                            cv2.putText(depth_display, f'Max: {cam.max_depth_value} mm',
                                        (10, 60), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (255, 255, 255), 2)
                            cv2.putText(depth_display, f'Mean: {int(cam.mean_depth_value)} mm',
                                        (10, 90), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (255, 255, 255), 2)
                            cv2.putText(depth_display, f'Valid: {cam.valid_pixel_count}',
                                        (10, 120), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (255, 255, 255), 2)

                        resized = cv2.resize(depth_display, None,
                                             fx=self.display_scale, fy=self.display_scale)
                        cv2.imshow(f'{cam.label} - Depth', resized)

                    if self.show_histogram and cam.depth_image is not None:
                        hist = self._create_histogram(cam.depth_image)
                        if hist is not None:
                            cv2.imshow(f'{cam.label} - Depth Histogram', hist)

                    if cam.color_image is not None:
                        resized = cv2.resize(cam.color_image, None,
                                             fx=self.display_scale, fy=self.display_scale)
                        cv2.imshow(f'{cam.label} - Color', resized)

            key = cv2.waitKey(33)
            if key == 27 or key == ord('q'):
                rclpy.shutdown()
                break
            elif key == ord('c'):
                self.colormap = (self.colormap + 1) % 4
                self.get_logger().info(f'Colormap changed to: {self.colormap}')
            elif key == ord('h'):
                self.show_histogram = not self.show_histogram
                self.get_logger().info(f'Histogram display: {"ON" if self.show_histogram else "OFF"}')
            elif key == ord('s'):
                self.show_statistics = not self.show_statistics
                self.get_logger().info(f'Statistics display: {"ON" if self.show_statistics else "OFF"}')

    def destroy_node(self):
        self.display_running = False
        if hasattr(self, 'display_thread') and self.display_thread.is_alive():
            self.display_thread.join(timeout=1.0)
        cv2.destroyAllWindows()
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = CameraDisplayNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: 11.camera_display/cpp/camera_display/src/camera_display_node.cpp
/**
 * @file camera_display_node.cpp
 * @brief ROS2 node for displaying RGB and depth images from head and waist cameras
 *
 * Simultaneously subscribes to both ob_camera_head and ob_camera_waist,
 * displaying color and depth images with colormap, histogram, and statistics.
 */

#include <memory>
#include <string>
#include <vector>
#include <thread>
#include <mutex>
#include <atomic>
#include <functional>
#include <map>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include <opencv2/opencv.hpp>

struct CameraData {
  std::string label;
  cv::Mat depth_image;
  cv::Mat colored_depth;
  cv::Mat color_image;
  std::mutex depth_mutex;
  std::mutex color_mutex;
  double min_depth_value = 0;
  double max_depth_value = 0;
  double mean_depth_value = 0;
  int valid_pixel_count = 0;
};

class CameraDisplayNode : public rclcpp::Node {
 public:
  CameraDisplayNode()
      : Node("camera_display_node"), display_running_(true) {

    this->declare_parameter("colormap", 2);
    this->declare_parameter("max_depth", 5000.0);
    this->declare_parameter("min_depth", 0.0);
    this->declare_parameter("display_scale", 0.5);
    this->declare_parameter("show_histogram", true);
    this->declare_parameter("show_statistics", true);
    this->declare_parameter("enable_head", true);
    this->declare_parameter("enable_waist", true);

    colormap_ = this->get_parameter("colormap").as_int();
    max_depth_ = this->get_parameter("max_depth").as_double();
    min_depth_ = this->get_parameter("min_depth").as_double();
    display_scale_ = this->get_parameter("display_scale").as_double();
    show_histogram_ = this->get_parameter("show_histogram").as_bool();
    show_statistics_ = this->get_parameter("show_statistics").as_bool();
    bool enable_head = this->get_parameter("enable_head").as_bool();
    bool enable_waist = this->get_parameter("enable_waist").as_bool();

    rclcpp::QoS qos(rclcpp::KeepLast(10));
    qos.reliable();
    qos.durability_volatile();

    if (enable_head) {
      setup_camera("ob_camera_head", "Head", qos);
    }
    if (enable_waist) {
      setup_camera("ob_camera_waist", "Waist", qos);
    }

    if (cameras_.empty()) {
      RCLCPP_ERROR(this->get_logger(),
                   "No camera enabled! Set enable_head or enable_waist to true.");
      return;
    }

    display_thread_ = std::thread(&CameraDisplayNode::display_loop, this);

    RCLCPP_INFO(this->get_logger(), "Camera Display Node initialized");
    for (auto& [ns, cam] : cameras_) {
      RCLCPP_INFO(this->get_logger(), "  %s camera:", cam->label.c_str());
      RCLCPP_INFO(this->get_logger(), "    Depth: %s/depth/image_raw", ns.c_str());
      RCLCPP_INFO(this->get_logger(), "    Color: %s/color/image_raw", ns.c_str());
    }
    RCLCPP_INFO(this->get_logger(), "  Colormap: %d", colormap_);
    RCLCPP_INFO(this->get_logger(), "  Depth range: %.0f - %.0f mm", min_depth_, max_depth_);
  }

  ~CameraDisplayNode() {
    display_running_ = false;
    if (display_thread_.joinable()) {
      display_thread_.join();
    }
    cv::destroyAllWindows();
  }

 private:
  void setup_camera(const std::string& ns, const std::string& label,
                     const rclcpp::QoS& qos) {
    auto cam = std::make_shared<CameraData>();
    cam->label = label;
    cameras_[ns] = cam;

    depth_subs_.push_back(
        this->create_subscription<sensor_msgs::msg::Image>(
            ns + "/depth/image_raw", qos,
            [this, cam](const sensor_msgs::msg::Image::SharedPtr msg) {
              depth_callback(msg, cam);
            }));

    color_subs_.push_back(
        this->create_subscription<sensor_msgs::msg::Image>(
            ns + "/color/image_raw", qos,
            [this, cam](const sensor_msgs::msg::Image::SharedPtr msg) {
              color_callback(msg, cam);
            }));
  }

  void depth_callback(const sensor_msgs::msg::Image::SharedPtr msg,
                       std::shared_ptr<CameraData> cam) {
    std::lock_guard<std::mutex> lock(cam->depth_mutex);

    if (msg->encoding == "16UC1" || msg->encoding == "mono16") {
      cam->depth_image = cv::Mat(msg->height, msg->width, CV_16UC1,
                                  const_cast<uint8_t*>(msg->data.data())).clone();
    } else if (msg->encoding == "32FC1") {
      cv::Mat float_depth(msg->height, msg->width, CV_32FC1,
                           const_cast<uint8_t*>(msg->data.data()));
      float_depth.convertTo(cam->depth_image, CV_16UC1, 1000.0);
    } else {
      RCLCPP_WARN(this->get_logger(), "Unsupported depth encoding: %s",
                  msg->encoding.c_str());
      return;
    }

    update_depth_statistics(cam->depth_image, *cam);
    apply_colormap(cam->depth_image, cam->colored_depth);
  }

  void color_callback(const sensor_msgs::msg::Image::SharedPtr msg,
                       std::shared_ptr<CameraData> cam) {
    std::lock_guard<std::mutex> lock(cam->color_mutex);

    cv::Mat temp;
    if (msg->encoding == "rgb8" || msg->encoding == "RGB8") {
      temp = cv::Mat(msg->height, msg->width, CV_8UC3,
                      const_cast<uint8_t*>(msg->data.data())).clone();
      cv::cvtColor(temp, cam->color_image, cv::COLOR_RGB2BGR);
    } else if (msg->encoding == "bgr8" || msg->encoding == "BGR8") {
      cam->color_image = cv::Mat(msg->height, msg->width, CV_8UC3,
                                  const_cast<uint8_t*>(msg->data.data())).clone();
    } else if (msg->encoding == "rgba8" || msg->encoding == "RGBA8") {
      temp = cv::Mat(msg->height, msg->width, CV_8UC4,
                      const_cast<uint8_t*>(msg->data.data())).clone();
      cv::cvtColor(temp, cam->color_image, cv::COLOR_RGBA2BGR);
    } else if (msg->encoding == "bgra8" || msg->encoding == "BGRA8") {
      temp = cv::Mat(msg->height, msg->width, CV_8UC4,
                      const_cast<uint8_t*>(msg->data.data())).clone();
      cv::cvtColor(temp, cam->color_image, cv::COLOR_BGRA2BGR);
    } else {
      RCLCPP_WARN(this->get_logger(), "Unsupported color encoding: %s",
                  msg->encoding.c_str());
    }
  }

  void apply_colormap(const cv::Mat& depth, cv::Mat& colored) {
    cv::Mat normalized;
    depth.convertTo(normalized, CV_8UC1,
                    255.0 / (max_depth_ - min_depth_),
                    -min_depth_ * 255.0 / (max_depth_ - min_depth_));

    switch (colormap_) {
      case 0:
        colored = normalized.clone();
        return;
      case 1:
        cv::applyColorMap(normalized, colored, cv::COLORMAP_JET);
        break;
      case 2:
        cv::applyColorMap(normalized, colored, cv::COLORMAP_RAINBOW);
        break;
      case 3:
        cv::applyColorMap(normalized, colored, cv::COLORMAP_TURBO);
        break;
      default:
        cv::applyColorMap(normalized, colored, cv::COLORMAP_JET);
        break;
    }

    cv::Mat mask = (depth == 0);
    colored.setTo(cv::Scalar(0, 0, 0), mask);
  }

  void update_depth_statistics(const cv::Mat& depth, CameraData& cam) {
    cv::Mat mask = (depth > 0);
    int count = cv::countNonZero(mask);
    if (count == 0) {
      cam.min_depth_value = 0;
      cam.max_depth_value = 0;
      cam.mean_depth_value = 0;
      cam.valid_pixel_count = 0;
      return;
    }

    cv::Mat non_zero;
    depth.copyTo(non_zero, mask);
    double min_val, max_val;
    cv::minMaxLoc(non_zero, &min_val, &max_val, nullptr, nullptr, mask);

    cam.min_depth_value = min_val;
    cam.max_depth_value = max_val;
    cam.mean_depth_value = cv::mean(depth, mask)[0];
    cam.valid_pixel_count = count;
  }

  cv::Mat create_histogram(const cv::Mat& depth) {
    int histSize = 256;
    float range[] = {static_cast<float>(min_depth_), static_cast<float>(max_depth_)};
    const float* histRange = {range};

    cv::Mat hist;
    cv::calcHist(&depth, 1, 0, cv::Mat(), hist, 1, &histSize, &histRange);
    cv::normalize(hist, hist, 0, 1, cv::NORM_MINMAX);

    int hist_w = 512, hist_h = 200;
    cv::Mat hist_image(hist_h, hist_w, CV_8UC3, cv::Scalar(0, 0, 0));

    int bin_w = cvRound(static_cast<double>(hist_w) / histSize);
    for (int i = 1; i < histSize; i++) {
      cv::line(hist_image,
               cv::Point(bin_w * (i - 1),
                          hist_h - cvRound(hist.at<float>(i - 1) * hist_h)),
               cv::Point(bin_w * i,
                          hist_h - cvRound(hist.at<float>(i) * hist_h)),
               cv::Scalar(255, 255, 255), 2);
    }

    std::stringstream ss;
    ss << "Range: " << static_cast<int>(min_depth_) << "-"
       << static_cast<int>(max_depth_) << " mm";
    cv::putText(hist_image, ss.str(), cv::Point(10, 20),
                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(255, 255, 255), 1);

    return hist_image;
  }

  void display_loop() {
    for (auto& [ns, cam] : cameras_) {
      cv::namedWindow(cam->label + " - Depth", cv::WINDOW_NORMAL);
      cv::namedWindow(cam->label + " - Color", cv::WINDOW_NORMAL);
    }

    while (display_running_ && rclcpp::ok()) {
      for (auto& [ns, cam] : cameras_) {
        {
          std::lock_guard<std::mutex> lock(cam->depth_mutex);
          if (!cam->colored_depth.empty()) {
            cv::Mat depth_display = cam->colored_depth.clone();

            if (show_statistics_) {
              auto put = [&](int y, const std::string& text) {
                cv::putText(depth_display, text, cv::Point(10, y),
                            cv::FONT_HERSHEY_SIMPLEX, 0.7,
                            cv::Scalar(255, 255, 255), 2);
              };
              put(30, "Min: " + std::to_string(static_cast<int>(cam->min_depth_value)) + " mm");
              put(60, "Max: " + std::to_string(static_cast<int>(cam->max_depth_value)) + " mm");
              put(90, "Mean: " + std::to_string(static_cast<int>(cam->mean_depth_value)) + " mm");
              put(120, "Valid: " + std::to_string(cam->valid_pixel_count));
            }

            cv::Mat resized;
            cv::resize(depth_display, resized, cv::Size(),
                       display_scale_, display_scale_);
            cv::imshow(cam->label + " - Depth", resized);
          }

          if (show_histogram_ && !cam->depth_image.empty()) {
            cv::Mat hist = create_histogram(cam->depth_image);
            if (!hist.empty()) {
              cv::imshow(cam->label + " - Depth Histogram", hist);
            }
          }
        }

        {
          std::lock_guard<std::mutex> lock(cam->color_mutex);
          if (!cam->color_image.empty()) {
            cv::Mat resized;
            cv::resize(cam->color_image, resized, cv::Size(),
                       display_scale_, display_scale_);
            cv::imshow(cam->label + " - Color", resized);
          }
        }
      }

      char key = cv::waitKey(33);
      if (key == 27 || key == 'q') {
        rclcpp::shutdown();
        break;
      } else if (key == 'c') {
        colormap_ = (colormap_ + 1) % 4;
        RCLCPP_INFO(this->get_logger(), "Colormap changed to: %d", colormap_);
      } else if (key == 'h') {
        show_histogram_ = !show_histogram_;
        RCLCPP_INFO(this->get_logger(), "Histogram display: %s",
                    show_histogram_ ? "ON" : "OFF");
      } else if (key == 's') {
        show_statistics_ = !show_statistics_;
        RCLCPP_INFO(this->get_logger(), "Statistics display: %s",
                    show_statistics_ ? "ON" : "OFF");
      }
    }
  }

  std::map<std::string, std::shared_ptr<CameraData>> cameras_;

  std::vector<rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr> depth_subs_;
  std::vector<rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr> color_subs_;

  int colormap_;
  double max_depth_;
  double min_depth_;
  double display_scale_;
  bool show_histogram_;
  bool show_statistics_;

  std::thread display_thread_;
  std::atomic<bool> display_running_;
};

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CameraDisplayNode>());
  rclcpp::shutdown();
  return 0;
}
